test/functional/wallet_avoid_mixing_output_types.py

```diff
@@ -0,0 +1,171 @@
+#!/usr/bin/env python3
+# Copyright (c) 2022 The Bitcoin Core developers
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+"""Test output type mixing during coin selection
+
+A wallet may have different type UTXOs to choose from during coin selection,
+where output type is one of the following:
+    - BECH32M
+    - BECH32
+    - P2SH-SEGWIT
+    - LEGACY
+
+This test verifies that mixing different output types is avoided unless
+absolutely necessary. Both wallets start with zero funds and Alice mines
+enough blocks to have spendable coinbase outputs. Alice sends a series of 10
+random value payments to Bob, creating 5 bech32 UTXOs and 5 P2SH UTXOs,
+for a total of 30BTC in Bob's wallet.
+
+Bob then sends random valued payments back to Alice, some of which need
+unconfirmed change, and we verify that none of these payments contain mixed
+inputs. Finally, Bob sends the remainder of his funds, which requires mixing.
+
+They payment values are random, but chosen such that they sum up to a specified
+total. This ensures we are not relying on specific values for the UTXOs,
+but still know when to expect mixing due to the wallet being close to empty.
+
+"""
+
+import random
+from test_framework.test_framework import BitcoinTestFramework
+from test_framework.blocktools import COINBASE_MATURITY
+
+
+def is_bech32_address(node, addr):
+    """Check if an address cointains a bech32 output."""
+    addr_info = node.getaddressinfo(addr)
+    if addr_info['iswitness']:
+        return (
+            not addr_info['isscript'] and
+            addr_info['iswitness'] and addr_info['witness_version'] == 0
+        )
+    return False
+
+
+def is_p2sh_address(node, addr):
+    """Check if an address cointains a P2SH output."""
+    addr_info = node.getaddressinfo(addr)
+    return (
+        addr_info['isscript'] and
+        not addr_info['iswitness']
+    )
+
+
+def is_legacy_address(node, addr):
+    """Check if an address cointains a legacy output."""
+    addr_info = node.getaddressinfo(addr)
+    return (not addr_info['isscript'] and not addr_info['iswitness'])
+
+
+def is_same_type(node, tx):
+    """Check that all inputs are of the same OutputType"""
+    vins = node.getrawtransaction(tx, True)['vin']
+    inputs = []
+    for vin in vins:
+        prev_tx, n = vin['txid'], vin['vout']
+        inputs.append(
+            node.getrawtransaction(
+                prev_tx,
+                True,
+            )['vout'][n]['scriptPubKey']['address']
+        )
+    has_legacy = False
+    has_p2sh = False
+    has_bech32 = False
+
+    for addr in inputs:
+        if is_legacy_address(node, addr):
+            has_legacy = True
+        if is_p2sh_address(node, addr):
+            has_p2sh = True
+        if is_bech32_address(node, addr):
+            has_bech32 = True
```

Comment:
I would have expected to count all five types (incl. bech32m and others), even if you don't explicitly create them here. This test would currently not fail if I Alice additionally sent Bob a P2TR output.

Reply:
good point, ill update this